superbench/benchmarks/micro_benchmarks/directx_gpu_copy_performance/GPUCopyBw.cpp

```diff
@@ -0,0 +1,249 @@
+    m_commandList->Close();
+    ID3D12CommandList *listsToExecute[] = {m_commandList.Get()};
+    this->m_commandQueue->ExecuteCommandLists(ARRAYSIZE(listsToExecute), listsToExecute);
+    // Signal and increment the fence value.
+    const UINT64 fenceL = m_copyFenceValue;
+    m_commandQueue->Signal(m_copyFence.Get(), fenceL);
+    m_copyFenceValue++;
+    // Wait until command queue is done.
+    if (m_copyFence->GetCompletedValue() < fenceL) {
+        m_copyFence->SetEventOnCompletion(fenceL, m_copyEventHandle);
+        WaitForSingleObject(m_copyEventHandle, INFINITE);
+    }
+    // Reset the command allocator and command list.
+    ID3D12CommandAllocator *activeAllocator = m_commandAllocator.Get();
+    activeAllocator->Reset();
+    m_commandList->Reset(activeAllocator, nullptr);
+}
+
+/**
+ * @brief Prepare data of the source buffer of benchmark.
+ * @param pData the data that should upload.
+ * @param byteSize the size of data.
+ */
+void GPUCopyBw::PrepareSourceBufferData(const void *pData, SIZE_T byteSize) {
+    // Upload data from CPU to upload buffer.
+    void *p;
+    m_uploadBuffer->Map(0, nullptr, &p);
+    memcpy(p, pData, byteSize);
+    m_uploadBuffer->Unmap(0, nullptr);
+
+    if (opts->dtoh_enabled) {
+        // Upload data from upload to default buffer.
+        CopyResourceFromUploadToDefault();
+        D3D12_RESOURCE_BARRIER outputBufferResourceBarrier{CD3DX12_RESOURCE_BARRIER::Transition(
+            m_vertexBuffer.Get(), D3D12_RESOURCE_STATE_COPY_DEST, D3D12_RESOURCE_STATE_COPY_SOURCE)};
+        m_commandList->ResourceBarrier(1, &outputBufferResourceBarrier);
+        ExecuteWaitForCopyQueue();
+    }
+}
+
+/**
+ * @brief Create pipeline including
+ *		  create device object, command list, command queue
+ *		  and synchronization objects.
+ */
+void GPUCopyBw::LoadPipeline() {
+    UINT dxgiFactoryFlags = 0;
+
+#if _DEBUG
+    // Enable the debug layer (requires the Graphics Tools "optional feature").
+    // NOTE: Enabling the debug layer after device creation will invalidate the active device.
+    {
+        ComPtr<ID3D12Debug> debugController;
+        if (SUCCEEDED(D3D12GetDebugInterface(IID_PPV_ARGS(&debugController)))) {
+            debugController->EnableDebugLayer();
+
+            // Enable additional debug layers.
+            dxgiFactoryFlags |= DXGI_CREATE_FACTORY_DEBUG;
+        }
+    }
+#endif
+
+    ComPtr<IDXGIFactory4> factory;
+    ThrowIfFailed(CreateDXGIFactory2(dxgiFactoryFlags, IID_PPV_ARGS(&factory)));
+
+    ComPtr<IDXGIAdapter1> hardwareAdapter;
+    GetHardwareAdapter(factory.Get(), &hardwareAdapter);
+
+    ThrowIfFailed(D3D12CreateDevice(hardwareAdapter.Get(), D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(&m_device)));
+
+    D3D12_COMMAND_QUEUE_DESC cqd3 = {};
+    cqd3.Type = D3D12_COMMAND_LIST_TYPE_COPY;
+    m_device->CreateCommandQueue(&cqd3, IID_PPV_ARGS(&m_commandQueue));
```

Comment:
>CreateCommandQueue [](http://example.com/codeflow?start=14&length=18)

actually functions like this returns HRESULT to show if it executes correctly. I guess you may check the return value to make sure it executes without any errors before go ahead.

Reply:
thanks, added throwiffailed to all functions which returns HRESULT